File: go/tetris.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"runtime"
	"strconv"
	"strings"
	"time"
)

const width = 10
const height = 20

var shapes = [][][]int{
	{{1, 1, 1, 1}},
	{{1, 1}, {1, 1}},
	{{0, 1, 0}, {1, 1, 1}},
	{{1, 0, 0}, {1, 1, 1}},
	{{0, 0, 1}, {1, 1, 1}},
	{{0, 1, 1}, {1, 1, 0}},
	{{1, 1, 0}, {0, 1, 1}},
}

type Tetris struct {
	field     [][]int
	score     int
	gameOver  bool
	piece     [][]int
	pieceX    int
	pieceY    int
}

func NewTetris() *Tetris {
	t := &Tetris{
		field: make([][]int, height),
		score: 0,
		gameOver: false,
	}
	for i := range t.field {
		t.field[i] = make([]int, width)
	}
	t.newPiece()
	return t
}

func (t *Tetris) newPiece() {
	idx := rand.Intn(len(shapes))
	t.piece = make([][]int, len(shapes[idx]))
	for i := range shapes[idx] {
		t.piece[i] = make([]int, len(shapes[idx][i]))
		copy(t.piece[i], shapes[idx][i])
	}
	t.pieceX = width/2 - len(t.piece[0])/2
	t.pieceY = 0
	if t.collision() {
		t.gameOver = true
	}
}

func (t *Tetris) collision() bool {
	for y, row := range t.piece {
		for x, cell := range row {
			if cell == 0 {
				continue
			}
			fx := t.pieceX + x
			fy := t.pieceY + y
			if fx < 0 || fx >= width || fy >= height || (fy >= 0 && t.field[fy][fx] != 0) {
				return true
			}
		}
	}
	return false
}

func (t *Tetris) merge() {
	for y, row := range t.piece {
		for x, cell := range row {
			if cell != 0 {
				t.field[t.pieceY+y][t.pieceX+x] = 1
			}
		}
	}
	t.clearLines()
	t.newPiece()
}

func (t *Tetris) clearLines() {
	lines := 0
	for y := height - 1; y >= 0; {
		full := true
		for x := 0; x < width; x++ {
			if t.field[y][x] == 0 {
				full = false
				break
			}
		}
		if full {
			// сдвиг вниз
			for i := y; i > 0; i-- {
				t.field[i] = t.field[i-1]
			}
			t.field[0] = make([]int, width)
			lines++
		} else {
			y--
		}
	}
	t.score += lines * 100
}

func (t *Tetris) move(dx, dy int) bool {
	t.pieceX += dx
	t.pieceY += dy
	if t.collision() {
		t.pieceX -= dx
		t.pieceY -= dy
		if dy == 1 {
			t.merge()
		}
		return false
	}
	return true
}

func (t *Tetris) rotate() {
	rows := len(t.piece)
	cols := len(t.piece[0])
	rotated := make([][]int, cols)
	for i := range rotated {
		rotated[i] = make([]int, rows)
	}
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			rotated[j][rows-1-i] = t.piece[i][j]
		}
	}
	oldPiece := t.piece
	t.piece = rotated
	if t.collision() {
		t.piece = oldPiece
	}
}

func clearScreen() {
	cmd := "clear"
	if runtime.GOOS == "windows" {
		cmd = "cls"
	}
	c := exec.Command(cmd)
	c.Stdout = os.Stdout
	c.Run()
}

func (t *Tetris) draw() {
	clearScreen()
	fmt.Println("Score:", t.score)
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			cell := t.field[y][x] != 0
			if !cell && t.pieceY <= y && y < t.pieceY+len(t.piece) &&
				t.pieceX <= x && x < t.pieceX+len(t.piece[0]) &&
				t.piece[y-t.pieceY][x-t.pieceX] != 0 {
				cell = true
			}
			if cell {
				fmt.Print("[]")
			} else {
				fmt.Print("  ")
			}
		}
		fmt.Println()
	}
	if t.gameOver {
		fmt.Println("GAME OVER")
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	game := NewTetris()
	reader := bufio.NewReader(os.Stdin)
	for !game.gameOver {
		game.draw()
		fmt.Print("> ")
		input, _ := reader.ReadString('\n')
		input = strings.TrimSpace(input)
		switch input {
		case "a":
			game.move(-1, 0)
		case "d":
			game.move(1, 0)
		case "s":
			game.move(0, 1)
		case "w":
			game.rotate()
		case "q":
			game.gameOver = true
		default:
			game.move(0, 1)
		}
	}
	game.draw()
	fmt.Println("Final score:", game.score)
}

File: python/tetris.py

### **2.2. Код Тетриса на 4 языках**

#### **Python (`python/tetris.py`) — консольная версия без внешних библиотек**

```python
import random
import os
import time

# Размеры поля
WIDTH = 10
HEIGHT = 20

# Фигурки тетрамино
SHAPES = [
    [[1, 1, 1, 1]],  # I
    [[1, 1], [1, 1]],  # O
    [[0, 1, 0], [1, 1, 1]],  # T
    [[1, 0, 0], [1, 1, 1]],  # L
    [[0, 0, 1], [1, 1, 1]],  # J
    [[0, 1, 1], [1, 1, 0]],  # S
    [[1, 1, 0], [0, 1, 1]]   # Z
]

class Tetris:
    def __init__(self):
        self.field = [[0] * WIDTH for _ in range(HEIGHT)]
        self.score = 0
        self.game_over = False
        self.new_piece()

    def new_piece(self):
        self.piece = random.choice(SHAPES)
        self.piece_x = WIDTH // 2 - len(self.piece[0]) // 2
        self.piece_y = 0
        if self.collision():
            self.game_over = True

    def collision(self):
        for y, row in enumerate(self.piece):
            for x, cell in enumerate(row):
                if cell:
                    field_x = self.piece_x + x
                    field_y = self.piece_y + y
                    if (field_x < 0 or field_x >= WIDTH or
                        field_y >= HEIGHT or
                        (field_y >= 0 and self.field[field_y][field_x])):
                        return True
        return False

    def merge(self):
        for y, row in enumerate(self.piece):
            for x, cell in enumerate(row):
                if cell:
                    self.field[self.piece_y + y][self.piece_x + x] = 1
        self.clear_lines()
        self.new_piece()

    def clear_lines(self):
        lines_cleared = 0
        y = HEIGHT - 1
        while y >= 0:
            if all(self.field[y]):
                del self.field[y]
                self.field.insert(0, [0] * WIDTH)
                lines_cleared += 1
            else:
                y -= 1
        self.score += lines_cleared * 100

    def move(self, dx, dy):
        self.piece_x += dx
        self.piece_y += dy
        if self.collision():
            self.piece_x -= dx
            self.piece_y -= dy
            if dy == 1:
                self.merge()
            return False
        return True

    def rotate(self):
        rotated = list(zip(*self.piece[::-1]))
        old_piece = self.piece
        self.piece = [list(row) for row in rotated]
        if self.collision():
            self.piece = old_piece

    def draw(self):
        os.system('clear')
        # Верхняя граница
        print('+' + '-' * (WIDTH * 2) + '+')
        for y in range(HEIGHT):
            line = '|'
            for x in range(WIDTH):
                if (self.piece_y <= y < self.piece_y + len(self.piece) and
                    self.piece_x <= x < self.piece_x + len(self.piece[0]) and
                    self.piece[y - self.piece_y][x - self.piece_x]):
                    line += '[]'
                elif self.field[y][x]:
                    line += '[]'
                else:
                    line += '  '
            line += '|'
            print(line)
        print('+' + '-' * (WIDTH * 2) + '+')
        print(f'Score: {self.score}')
        if self.game_over:
            print('GAME OVER')

def main():
    game = Tetris()
    while not game.game_over:
        game.draw()
        cmd = input('> ')
        if cmd == 'a':
            game.move(-1, 0)
        elif cmd == 'd':
            game.move(1, 0)
        elif cmd == 's':
            game.move(0, 1)
        elif cmd == 'w':
            game.rotate()
        elif cmd == 'q':
            break
        else:
            game.move(0, 1)

if __name__ == '__main__':
    main()
